packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -18,34 +18,87 @@ import URI from '@theia/core/lib/common/uri';
             const currentLocation = this.service.location;
             locationList.value = currentLocation ? currentLocation.toString() : '';
+        } else if (locationListTextInput) {
+            setTimeout(() => locationListTextInput.focus());
         }
-    }
+    };
 
     protected readonly handleLocationChanged = (e: React.ChangeEvent<HTMLSelectElement>) => this.onLocationChanged(e);
+    protected readonly handleLocationKeydown = (e: React.KeyboardEvent<HTMLSelectElement>) => this.onLocationKeydown(e);
+    protected readonly handleTextInputOnChange = (e: React.ChangeEvent<HTMLInputElement>) => this.onTextInputChanged(e);
+    protected readonly handleTextInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => this.onTextInputKeyDown(e);
+    protected readonly handleTextInputOnBlur = () => this.onTextInputToggle(false);
+    protected readonly handleTextInputMouseDown = (e: React.MouseEvent<HTMLSpanElement>) => this.onTextInputToggle(e.currentTarget.id === 'select-input');
+
     protected doRender(): React.ReactNode {
         const options = this.collectLocations().map(value => this.renderLocation(value));
-        return <select className={'theia-select ' + LocationListRenderer.Styles.LOCATION_LIST_CLASS} onChange={this.handleLocationChanged}>{...options}</select>;
+        return (
+            <>
+                {!!this.fileService && <span onMouseDown={this.handleTextInputMouseDown}
+                    className={LocationListRenderer.Styles.LOCATION_INPUT_TOGGLE_CLASS}
+                    tabIndex={0}
+                    id={`${this.doShowTextInput ? 'text-input' : 'select-input'}`}
+                    title={this.doShowTextInput
+                        ? LocationListRenderer.Tooltips.TOGGLE_SELECT_INPUT
+                        : LocationListRenderer.Tooltips.TOGGLE_TEXT_INPUT}
+                >
+                    <i className={this.doShowTextInput ? 'fa fa-folder-open' : 'fa fa-edit'} />
+                </span>}
+                { this.doShowTextInput ?
+                    <input className={'theia-select ' + LocationListRenderer.Styles.LOCATION_TEXT_INPUT_CLASS}
+                        defaultValue={this.service.location?.path.toString()}
+                        onBlur={this.handleTextInputOnBlur}
+                        onChange={this.handleTextInputOnChange}
+                        onKeyDown={this.handleTextInputKeyDown}
+                        spellCheck={false}
+                    />
+                    :
+                    <select className={`theia-select ${LocationListRenderer.Styles.LOCATION_LIST_CLASS} ${!!this.fileService ? 'with-icon' : ''}`}
+                        onKeyDown={this.handleLocationKeydown}
+                        onChange={this.handleLocationChanged}>F
+                        {...options}
+                    </select>
+                }
+            </>
+        );
     }
 
+    protected onTextInputToggle(shouldShowTextInput: boolean): void {
```

Comment:
The `on` here might lead one to believe that it's an event rather than an event handler.

Reply:
Fair point, I was following the existing convention but agree with what you're saying. I've also updated the other new handlers to have more description method names